Forward AD formulas batch 3

albanD opened the pull request:
Slow gradcheck also passes for this PR and can be found here: https://github.com/pytorch/pytorch/pull/57976

Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#58094 Forward AD formulas batch 3**
* #59448 ger is an alias to outer, not the other way around

Differential Revision: [D28387762](https://our.internmc.facebook.com/intern/diff/D28387762)

albanD committed:
Forward AD formulas batch 3

[ghstack-poisoned]

albanD committed:
Update on "Forward AD formulas batch 3"




[ghstack-poisoned]

albanD committed:
Update on "Forward AD formulas batch 3"




[ghstack-poisoned]

albanD commented:
@albanD has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28387762).

albanD committed:
Update on "Forward AD formulas batch 3"

Slow gradcheck also passes for this PR and can be found here: https://github.com/pytorch/pytorch/pull/57976


Differential Revision: [D28387762](https://our.internmc.facebook.com/intern/diff/D28387762)

[ghstack-poisoned]

albanD committed:
Update on "Forward AD formulas batch 3"

Slow gradcheck also passes for this PR and can be found here: https://github.com/pytorch/pytorch/pull/57976


Differential Revision: [D28387762](https://our.internmc.facebook.com/intern/diff/D28387762)

[ghstack-poisoned]

albanD committed:
Update on "Forward AD formulas batch 3"

Slow gradcheck also passes for this PR and can be found here: https://github.com/pytorch/pytorch/pull/57976


Differential Revision: [D28387762](https://our.internmc.facebook.com/intern/diff/D28387762)

[ghstack-poisoned]

albanD committed:
Update on "Forward AD formulas batch 3"

Slow gradcheck also passes for this PR and can be found here: https://github.com/pytorch/pytorch/pull/57976


Differential Revision: [D28387762](https://our.internmc.facebook.com/intern/diff/D28387762)

[ghstack-poisoned]

albanD commented:
@albanD has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28387762).

albanD committed:
Update on "Forward AD formulas batch 3"

Slow gradcheck also passes for this PR and can be found here: https://github.com/pytorch/pytorch/pull/57976


Differential Revision: [D28387762](https://our.internmc.facebook.com/intern/diff/D28387762)

[ghstack-poisoned]

zou3519 commented on tools/autograd/derivatives.yaml:
```diff
@@ -431,9 +431,11 @@
 
 - name: cummax(Tensor self, int dim) -> (Tensor values, Tensor indices)
```
(note to self): here's a list of all the formulas in this diff: https://gist.github.com/zou3519/f5e818f8586854f6d018e52398920fa8

zou3519 commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -5313,6 +5351,7 @@ def gradcheck_wrapper_triangular_input(op, input, *args, upper=False, **kwargs):
            dtypesIfCPU=all_types_and_complex_and(torch.float16, torch.bfloat16),
            dtypesIfCUDA=floating_and_complex_types_and(torch.float16, *[torch.bfloat16] if CUDA11OrLater else []),
            assert_autodiffed=True,
+           supports_forward_ad=True,
```
This should be in the last PR

albanD replied:
No, the formula for "mm" is added in this PR line 930 of derivatives.yaml

zou3519 replied:
thanks for pointing that out, I missed it

zou3519 commented on tools/autograd/derivatives.yaml:
```diff
@@ -443,6 +445,7 @@
 
 - name: deg2rad(Tensor self) -> Tensor
   self: deg2rad_backward(grad)
+  result: auto_element_wise
```
It's probably too late for this, but in the future, if there are more auto_element_wise and auto_linear formulas, if you separate them out from the formulas that need more thinking about we can review and ship them faster

albanD replied:
Sure!

albanD committed:
Update on "Forward AD formulas batch 3"

Slow gradcheck also passes for this PR and can be found here: https://github.com/pytorch/pytorch/pull/57976


Differential Revision: [D28387762](https://our.internmc.facebook.com/intern/diff/D28387762)

[ghstack-poisoned]

albanD commented:
@albanD has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28387762).

albanD commented:
@albanD has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28387762).

zou3519 commented on tools/autograd/derivatives.yaml:
```diff
@@ -463,24 +466,30 @@
 - name: div.Tensor_mode(Tensor self, Tensor other, *, str? rounding_mode) -> Tensor
   self: div_tensor_self_backward(grad, other, self.scalar_type(), rounding_mode)
   other: div_tensor_other_backward(grad, self, other, rounding_mode)
+  result: "rounding_mode.has_value() ? result.new_zeros(result.sizes()) : self_t / other_p - other_t * (self_p / other_p) / other_p"
```
result.new_zeros(result.sizes()) -> at::zeros_like(result) ?

zou3519 commented on tools/autograd/derivatives.yaml:
```diff
@@ -574,6 +591,7 @@
 - name: ger(Tensor self, Tensor vec2) -> Tensor
   self: grad.mv(vec2.conj())
   vec2: grad.t().mv(self.conj())
+  result: at::ger(self_t, vec2_p) + at::ger(self_p, vec2_t)
```
I don't see a test for ger, should it go here? https://github.com/pytorch/pytorch/blob/d1638820935455b4c7d2c6c60baf46827a1bc9a8/torch/testing/_internal/common_methods_invocations.py#L5605-L5609

albanD replied:
outer actually isn't an alias at all...
It is ger that calls into outer. But weirdly enough, ger has a backward formula while outer is a composite op.

@mruberry do you know why these two ops that are supposed to be alias don't have the same backward definition (one has an explicit formula and the other is composite)?

mruberry replied:
That is weird! Almost definitely an oversight. I would expect `ger` to be a proper alias for `outer`.

albanD replied:
Ok, sending a PR to fix this :)

zou3519 commented on torch/csrc/autograd/FunctionsManual.cpp:
```diff
@@ -3535,6 +3542,18 @@ Tensor cumprod_jvp(Tensor self_t, Tensor self_p, Tensor result, int dim) {
   }
 }
 
+Tensor gather_with_keepdimed_indices(const Tensor& input, int64_t dim, const Tensor& indices, bool keepdim) {
```
I like how keepdimed is an adjective

zou3519 commented on tools/autograd/derivatives.yaml:
```diff
@@ -649,27 +670,33 @@
   index: non_differentiable
+  result: self_t.index_fill_(dim, index, value_t)
 
 - name: index_put_(Tensor(a!) self, Tensor?[] indices, Tensor values, bool accumulate=False) -> Tensor(a!)
   self: "accumulate ? grad : grad.clone().index_put_(indices, zeros_like(values), false)"
   values: grad.index(indices)
+  result: self_t.index_put_(indices, values_t, accumulate)
 
 - name: _index_put_impl_(Tensor(a!) self, Tensor?[] indices, Tensor values, bool accumulate=False, bool unsafe=False) -> Tensor(a!)
   self: "accumulate ? grad : grad.clone().index_put_(indices, zeros_like(values), false)"
   values: grad.index(indices)
+  result: at::_index_put_impl_(self_t, indices, values_t, accumulate, unsafe)
```
Why are index_put_ and _index_put_impl_ different?

albanD replied:
Not sure to be honest... It does look weird. I'm not even sure why the `impl` version exists.. Is that ok if I file this as a follow up task?

zou3519 replied:
Yeah, feel free to handle this in a follow-up task. The formula looks correct to me, so I don't think anything (aside from test coverage) could go wrong

albanD replied:
After looking into more details, I think the reason for this is that an extra argument was needed (unsafe) but to avoid exposing it externally, the impl version was added.

zou3519 commented on tools/autograd/derivatives.yaml:
```diff
@@ -693,63 +721,81 @@
 - name: xlogy.Tensor(Tensor self, Tensor other) -> Tensor
   self: grad * at::xlogy((self != 0), other)
   other: grad * self / other
+  result: self_t * at::xlogy((self_p != 0), other_p) + other_t * self_p / other_p
```
(no action required) Lots of binary elementwise formulas here that we could have codegen'ed with a special "auto_element_wise"

zou3519 reviewed (commented):
Went through the formulas, they all look correct to me. My questions are only around where the tests for `ger` and `_index_put_impl_` are

albanD committed:
Update on "Forward AD formulas batch 3"

Slow gradcheck also passes for this PR and can be found here: https://github.com/pytorch/pytorch/pull/57976


Differential Revision: [D28387762](https://our.internmc.facebook.com/intern/diff/D28387762)

[ghstack-poisoned]

albanD commented:
@zou3519 this should be good to go now!
I moved the ger/outer update to a PR below to keep things simple.

albanD commented:
@albanD has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28387762).

albanD committed:
Update on "Forward AD formulas batch 3"

Slow gradcheck also passes for this PR and can be found here: https://github.com/pytorch/pytorch/pull/57976


Differential Revision: [D28387762](https://our.internmc.facebook.com/intern/diff/D28387762)

[ghstack-poisoned]

janeyx99 commented:
Reverting as this PR broke some forward_mode_AD tests, e.g., https://app.circleci.com/pipelines/github/pytorch/pytorch/332766/workflows/901123bb-fbd5-48ef-b654-8e6926929c8f/jobs/13988206
